File: 2019-02/206-reverse-linked-list/js/206.js
#!/usr/bin/env node

let ListNode = (val) => ({
    val: val,
    next: null,
});

let printList = (root) => {
    if (root == null) {
        return;
    }
    console.log(root.val);
    printList(root.next);
};

let reverseList = (head) => reverseListRecur(head);

let reverseListIter = (head) => {
    let newList = ListNode(head.val);    
    let node = head.next;
    while (node != null) {
        let temp = ListNode(node.val);
        temp.next = newList;
        newList = temp;
        node = node.next;
    }
    return newList;
};

let reverseListRecur = (head) => head == null ? null : reverseListRecur_(head.next, ListNode(head.val));

let reverseListRecur_ = (remainingList, newList) => {
    if (remainingList == null) {
        return newList;
    }
    let temp = ListNode(remainingList.val); // 2
    temp.next = newList;
    newList = temp;
    return reverseListRecur_(remainingList.next, newList);
};

let input = ListNode(1);
input.next = ListNode(2);
input.next.next = ListNode(3);
input.next.next.next = ListNode(4);
input.next.next.next.next = ListNode(5);

console.log("Input list: ");
printList(input);
console.log("Output list: ");
printList( reverseList( input )) ;

File: 2020-01/206.py
#!/usr/bin/env python

# from typing import ListNode

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head

        prev = None
        current = head
        while current is not None:
            next = current.next
            current.next = prev
            prev = current
            current = next
        head = prev
        return head

        # first = head
        # #  1 -> 2 -> 3 -> None
        # #  ^    ^    ^
        # #  i    j    temp

        # #  1 <- 2 -> 3 -> None
        # #       ^    ^    ^
        # #       i    j    temp
        # i = head
        # j = head.next
        # temp = head.next.next

        # while temp is not None:
        #     j.next = i
        #     i = j
        #     j = temp
        #     temp = j.next

        # first.next = None
        # j.next = i
        # return j


def pl(l: ListNode):
    if l is None:
        return
    while l is not None:
        print(l.val, end=" ")
        l = l.next
    print("")


if __name__ == "__main__":
    l = ListNode(1)
    l.next = ListNode(2)
    l.next.next = ListNode(3)
    pl(l)

    S = Solution()
    rl = S.reverseList(l)
    pl(rl)
